Final improvements to Dynamo ML data pipeline.

reddyashish opened the pull request:
### Purpose

This addresses some final changes to the Dynamo ML pipeline and some comments made on [PR1](https://github.com/DynamoDS/Dynamo/pull/14749) and [PR2](https://github.com/DynamoDS/Dynamo/pull/15010).

Main changes:
1) User id, hostname and dynamo version are added to the request parameter.
2) The workspace log is sent to ML backend, when a saved workspace is closed instead of on Save(). This post restsharp task is run a different thread to avoid closing delays.
3) Checksum serialization is now using JSON.
4) Added exception handling. This should also handle 'no-network' mode.

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes
Final improvements to Dynamo ML data pipeline.

### Reviewers
@varvaratou @QilongTang @mjkkirschner

reddyashish commented on src/DynamoCore/Configuration/GraphChecksumItem.cs:
```diff
@@ -13,14 +12,4 @@ public class GraphChecksumItem
-    /// <summary>
-    /// Represents the stringified version of the nodes connections from a graph
-    /// </summary>
-    public class GraphChecksumPair
```
Was added after 3.0.3 release, so safe to remove this public API.

varvaratou commented on src/DynamoMLDataPipeline/DynamoMLDataPipeline.cs:
```diff
@@ -269,70 +282,20 @@ static void DataExchangeToForge(string filePath, RestClient client, string token
-        private static string GetAuthorizationToken()
-        {
-            return AuthTokenProvider.GetAccessToken();
+            SendToMLDataPipeline(filePath, client, token);
```
Just a reminder for the try catch block here.

reddyashish replied:
The main function calling this has a try catch, but will add it here as well.

reddyashish replied:
Added.

varvaratou commented on src/DynamoMLDataPipeline/DynamoMLDataPipeline.cs:
```diff
@@ -184,7 +197,7 @@ static void DataExchangeToForge(string filePath, RestClient client, string token
             ExchangeContainerId = exchangeRequestResponseBody["id"].Value;
 
-            Analytics.TrackEvent(Actions.Export, Categories.DynamoMLDataPipelineOperations, "ExchangeContainerID", Convert.ToInt32(ExchangeContainerId));
+            Analytics.TrackEvent(Actions.Export, Categories.DynamoMLDataPipelineOperations, ExchangeContainerId);
```
Is there a reason we send the data (CollectionId, ExchangeContainerId, BinarySize) as different Analytics events? Would it be possible to combine them into one Analytics log given that they are all part of the same export event? Analytics.NET seems to have an Analytics.TrackEvent call that accepts a dictionary of values. 
Ideally, on the ADP side we would like to be able to map CollectionId to ExchangeContainerId or monitor the BinarySize for a given collection.

reddyashish replied:
Yes changed it to a dictionary.

QilongTang commented on src/DynamoCore/Core/IDSDKManager.cs:
```diff
@@ -78,6 +78,18 @@ public string Username
+            get
+            {
+                var result = IDSDK_GetUserInfo();
+                return result != null ? result.UserId : String.Empty;
```
The userId currently is email?

mjkkirschner replied:
so it's the oxygen ID?

reddyashish replied:
yes it is oxygen id.

varvaratou commented on src/DynamoMLDataPipeline/DynamoMLDataPipeline.cs:
```diff
@@ -147,12 +167,8 @@ static public void EndFullFillment(RestClient client, string collectionId, strin
 
```
This comment is for lines 154-165: I know that it is outside the scope of the changes, but would it make sense to completely remove the while block? This was serving for testing, to confirm if the fulfillment was successfully completed. Since there is no follow up action dependent on the status of fulfillment why do we keep it? It would make sense to keep it if we want to keep track of successful versus not fulfillments, for sending an Analytics log for example. 
And even in that case, if something goes wrong on the Forge side and the fulfillment never succeeds, we will get stuck in the while loop.

reddyashish replied:
Yes removed the while loop.

mjkkirschner commented on src/DynamoCore/Core/IDSDKManager.cs:
```diff
@@ -78,6 +78,18 @@ public string Username
+            get
+            {
+                var result = IDSDK_GetUserInfo();
+                return result != null ? result.UserId : String.Empty;
```
if the user id is an empty string or null - should we still be sending data? How will the backend interpret that?

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:
```diff
@@ -776,23 +776,23 @@ protected DynamoViewModel(StartConfiguration startConfiguration)
+                    catch (Exception ex)
                     {
-                        checksums = (List<GraphChecksumPair>)serializer.Deserialize(reader);
+                        Model.Logger.Log("Failed to deserialize " + dynamoMLDataFileName + "::" +  ex.Message, LogLevel.File);
```
please use string interpolation with `${}`

mjkkirschner commented on src/DynamoMLDataPipeline/BinaryAsset.cs:
```diff
@@ -37,7 +37,7 @@ public BinaryAsset()
-    class BinaryAssets
+    internal class BinaryAssets
     {
         [JsonProperty("binaries")]
```
I think all these JsonProperties can be removed if you use serializer settings that set names to lower case.
https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/customize-properties?pivots=dotnet-8-0

less code to maintain is better.

reddyashish replied:
Will address this in the next version, as more changes are needed and additional testing would be needed.

mjkkirschner commented on src/DynamoMLDataPipeline/DynamoMLDataPipeline.cs:
```diff
@@ -49,34 +49,53 @@ internal string ProductionCollectionID
+        {
+            get
+            {
+                return AuthProvider.UserId;
```
Have you tried this in Revit? You have not added a new interface, and implemented it to make sure that ALL of of the auth providers actually have this method.

In Revit, `IDSDKManager` type will not be loaded, that's only in sandbox.

reddyashish replied:
Yes, this was one of the questions I had before testing on Revit. I will look into this.

reddyashish replied:
Works fine in Revit.

mjkkirschner commented on src/DynamoMLDataPipeline/DynamoMLDataPipeline.cs:
```diff
@@ -49,34 +49,53 @@ internal string ProductionCollectionID
-        internal static IOAuth2AccessTokenProvider AuthTokenProvider { get; set; }
+        internal DynamoModel DynamoModel { get; set; }
+
+        internal static IDSDKManager AuthProvider { get; set; }
```
I'd think hard about making this static...

reddyashish replied:
Removed static modifier.

mjkkirschner commented on src/DynamoMLDataPipeline/DynamoMLDataPipelineExtension.cs:
```diff
@@ -24,15 +20,10 @@ public string UniqueId
         {
             DynamoMLDataPipeline = new DynamoMLDataPipeline();
-            DynamoMLDataPipeline.AuthTokenProvider = (IOAuth2AccessTokenProvider)sp.AuthProvider;
+            DynamoMLDataPipeline.AuthProvider = (IDSDKManager)sp.AuthProvider;
```
yeah so, this will throw in Revit and Civil.

reddyashish replied:
Using IOAuth2UserIDProvider interface now, will update the Revit PR to extend this interface.

varvaratou commented on src/DynamoMLDataPipeline/DynamoMLDataPipeline.cs:
```diff
@@ -219,7 +220,7 @@ static void DataExchangeToForge(string filePath, RestClient client, string token
             uploadBinaryRequest.AddHeader("Content-Type", ContentType.Binary);
 
-            uploadBinaryRequest.AddOrUpdateParameter(ContentType.Binary.ToString(), base64CompressedBuffer);
+            uploadBinaryRequest.AddOrUpdateParameter(ContentType.Binary.ToString(), base64CompressedBuffer, ParameterType.RequestBody);
```
@reddyashish I have been testing the pipeline on the current status of the branch - if we switch the content type to `text/txt` in this line, we solve the problem with getting the unwanted headers in the response body when we retrieve back the binary. After that the conversion to .dyn works as expected:
`uploadBinaryRequest.AddOrUpdateParameter("text/txt", base64CompressedBuffer, ParameterType.RequestBody)`

reddyashish replied:
sounds good, changed it.

reddyashish commented on src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:
```diff
@@ -209,7 +209,7 @@ internal bool IsMLDataIngestionPipelineinBeta
             get
             {
-                return DynamoModel.FeatureFlags?.CheckFeatureFlag("IsMLDataIngestionPipelineinBeta", false) ?? false;
+                return DynamoModel.FeatureFlags?.CheckFeatureFlag("IsMLDataIngestionPipelineinBeta", true) ?? true;
```
@QilongTang Confirm if the default value for this flag should be true and in the case feature flags object is null.

reddyashish commented:
Addressed all the comments here, please take a final look before we can merge. Testing in Revit to make sure everything is good.
I will open a PR to GregOAuth repo to handle interface changes for revit host.

QilongTang commented on src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:
```diff
@@ -2936,6 +2919,32 @@ public void MakeNewHomeWorkspace(object parameter)
+            // Upon closing a workspace, validate if the workspace is valid to be sent to the ML datapipeline and then send it.
+            if (!DynamoModel.IsTestMode && !HomeSpace.HasUnsavedChanges && (currentWorkspaceViewModel?.IsHomeSpace ?? true) && HomeSpace.HasRunWithoutCrash)
+            {
+                if (!IsMLDataIngestionPipelineinBeta && Model.CurrentWorkspace.IsValidForFDX && currentWorkspaceViewModel.Checksum != string.Empty)
```
This would be disabled by default correct with the value being `true`?

reddyashish replied:
Yes will be disabled with the flag value true. We will be switching the flag value to false once the release is done?

QilongTang replied:
@reddyashish Yes, once we know it is integrated into Revit for sure

reddyashish commented:
Revit testing worked as expected. The only thing is userid is empty for now in Revit because the RevitAuthProvider is yet to implement the userid provider interface. That should be fixed after the PR is merged to GregOAuth repo.

reddyashish commented:
PR to handle RevitAuthProvider https://git.autodesk.com/Dynamo/GRegOAuth/pull/32

QilongTang commented:
Looks like there is no resource change in this PR which is good.

reddyashish commented:
@QilongTang @varvaratou Removed some of the patterns used for PII removal, so that the workspace is not altered. This should be the final changes now.

QilongTang commented:
There is one regression from the last change I think [Dynamo.Tests.WorkspaceSaving.RemovePIIDataFromWorkspace](https://master-15.jenkins.autodesk.com/job/DYN-Dynamo/job/DynamoBuildscripts/job/master/2662/testReport/junit/Dynamo.Tests/WorkspaceSaving/RemovePIIDataFromWorkspace/). Since this test is currently disabled on RC branch so RC branch is not affected
